Support nullable

mdmower-csnw opened the pull request:
- Add transform for `nullable: true` to include `null` as a possible type for any of the following schema types: `const`, `enum`, `type`.
- Log a warning when `const` is not set to `null` but schema is marked `nullable: true`
- Log a warning when `enum` does not contain `null` but schema is marked `nullable: true`
- Add tests and update snapshot

References from upstream:
- #522
- #535

mdmower-csnw committed:
Support nullable

- Add transform for nullable: true to include null as a possible type
  for any of the following schema types: const, enum, type, anyOf, and
  oneOf.
- Log a warning when enum does not contain null but schema is marked
  nullable: true.
- Add tests and update snapshot

tvharris commented on src/normalizer.ts:
```diff
@@ -222,6 +223,32 @@ rules.set('Transform const to singleton enum', schema => {
+
+  delete schema.nullable
+
+  if (schema.const !== undefined) {
+    if (schema.const !== null) {
+      schema.enum = [schema.const, null]
+      delete schema.const
+    }
+  } else if (schema.enum) {
+    if (!schema.enum.includes(null)) {
+      schema.enum.push(null)
+      log('yellow', 'normalizer', 'enum should include null when schema is nullable', schema)
```
@mdmower-csnw: 
1. Should there be a warning for the `const` case, too? [AJV](https://ajv.js.org/json-schema.html#nullable) docs say with `nullable: true`:
> const would fail, unless it is "const": null

2. I wouldn't be opposed to throwing instead of the warnings, because it's critical that the schema is updated or the type will be at odds with how AJV validates.

3. What if `const` and `enum` are both defined? Is that case caught earlier by validation?

mdmower-csnw replied:
The `const` case is mostly moot because normalizer _Transform const to singleton enum_ runs just before _Transform nullable to null type_:

```ts
rules.set('Transform const to singleton enum', schema => {
  if (schema.const !== undefined) {
    schema.enum = [schema.const]
    delete schema.const
  }
})
```

I included `const` handling just in case normalizers are reordered in the future.

> What if const and enum are both defined? Is that case caught earlier by validation?

`schema.enum` actually gets wiped out in this case due to the above normalizer. This is a great example of how this tool is not a validator. There's an expectation that the user has performed validation prior to using this tool.

> Should there be a warning for the const case, too? [AJV](https://ajv.js.org/json-schema.html#nullable) docs say with `nullable: true`.

That's fine. Can do.

> I wouldn't be opposed to throwing instead of the warnings, because it's critical that the schema is updated or the type will be at odds with how AJV validates.

Like mentioned before, this is not a validator. It's a tool that seldom throws.

mdmower-csnw replied:
Opted to reorder transforms so that _Transform nullable to null type_ runs before _Transform const to singleton enum_ in [a5db6da](https://github.com/CSNW/json-schema-to-typescript/pull/1/commits/a5db6da52b5992917f08f911994d8ed2df58f1b5). Also created a `warning()` function that prints to the console even when `VERBOSE` is not used.

tvharris commented on src/normalizer.ts:
```diff
@@ -222,6 +223,32 @@ rules.set('Transform const to singleton enum', schema => {
+    }
+  } else if (schema.type) {
+    schema.type = [...[schema.type].flatMap(value => value), 'null']
+  } else if (schema.anyOf) {
+    schema.anyOf.push(link({type: 'null'}, schema.anyOf))
+  } else if (schema.oneOf) {
+    schema.oneOf.push(link({type: 'null'}, schema.oneOf))
```
@mdmower-csnw: Is this how AJV handles `nullable: true` for `anyOf` and `oneOf`? It doesn't say explicitly in the [docs](https://ajv.js.org/json-schema.html#nullable).

mdmower-csnw replied:
That is a rabbit hole.

The OpenAPI spec writers (i.e. not the full picture, just one example of a team defining a JSON schema) eventually decided to restrict use of `nullable` to only schemas that define `type` in 3.0.3 (see [3.0.3: Fields](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#fixed-fields-20)). In an early draft of 3.1, when `nullable` was planned for deprecation rather than removal, wording implied that `nullable` wasn't forbidden when paired with something like `allOf`, but that it could be overridden (see [3.1(draft): Fields](https://github.com/philsturgeon/OpenAPI-Specification/blob/bb9a6f273389d25c3dbbb1d126b1a6fa1d8a4c85/versions/3.1.0.md#fixed-fields-20)). The start of the rabbit hole is here: [OpenAPI-Specification: Reference objects don't combine well with “nullable”](https://github.com/OAI/OpenAPI-Specification/issues/1368).

I approached this from the standpoint that `nullable` paired with `anyOf` and `oneOf` is unambiguous (but not `allOf`). After more reading, it looks like this was a bad assumption and there's good reason the OpenAPI team opted to limit use of `nullable` to schemas with `type`. The most helpful comments were this inqury: https://github.com/OAI/OpenAPI-Specification/issues/1368#issuecomment-488678380, and this response: https://github.com/OAI/OpenAPI-Specification/issues/1368#issuecomment-488697690 .

I lean towards leaving the `const` and `enum` pairings with `nullable` as is, i.e. allowed, but logged if `null` not in value(s) -- the intent is clear even if not 100% spec compliant. On the other hand, I no longer feel comfortable pairing `nullable` with `anyOf`, `oneOf`, and `allOf`. These will be removed.

mdmower-csnw replied:
Handling of `anyOf` and `oneOf` removed in [a5db6da](https://github.com/CSNW/json-schema-to-typescript/pull/1/commits/a5db6da52b5992917f08f911994d8ed2df58f1b5).

tvharris commented on test/e2e/nullable.ts:
```diff
@@ -0,0 +1,74 @@
+      const: '',
+      nullable: true,
+    },
+    h: {
+      type: 'string',
+      const: null,
+      nullable: true,
+    },
```
@mdmower-csnw: This might not be directly related to the changes in this PR, but I was wondering if there's a warning for a schema like this where the `type` and `const` (or `enum` values) are inconsistent. I don't know how AJV handles such a schema.

mdmower-csnw replied:
There is not a warning output by this tool. You can test this.

tvharris commented on test/normalizer/nullableAddsNullToType.json:
```diff
@@ -0,0 +1,139 @@
+      "e": {
+        "type": ["string", "null"]
+      },
+      "f": {
+        "type": ["null", "null"]
+      },
```
@mdmower-csnw: The output here and cases `l`, `n`, `p` don't seem ideal with the redundant `null`. I suppose it doesn't matter since the types are generated correctly.

mdmower-csnw replied:
Right. I don't think there's need for deduplication of types at this stage since the end user doesn't see it and the TypeScript conversion handles duplicates just fine.

mdmower-csnw replied:
Deduplication was trivial so opted to handle it in [a5db6da](https://github.com/CSNW/json-schema-to-typescript/pull/1/commits/a5db6da52b5992917f08f911994d8ed2df58f1b5).

tvharris reviewed (approved):
@mdmower-csnw: Approved. This is going to be a huge help for us in multiple projects!

There are a few points to consider below. I reviewed with the mindset that the implementation should be consistent with AJV's because EOV uses AJV.

mdmower-csnw reviewed (commented):
> I reviewed with the mindset that the implementation should be consistent with AJV's because EOV uses AJV.

I wouldn't call that the guiding principle. This project is not affiliated with express-openapi-validator. Also, AJV is a great resource, but this tool is not a validator, so it doesn't have to be as strict. When I search the module for `throw `, it seems to me that the maintainer opts to work with what is available to the extent possible rather than throw because of minor validation issues. It's not my intention to start diverging significantly from upstream. If we can achieve our needs with a light touch, that's how I'd like to proceed.

mdmower-csnw committed:
(Revisions) Support nullable

- Introduce warning() to output console warnings
- Add warning for nullable const, where const value is not null
- Move nullable transform before const->enum transform so that potential
  warning about nullable const can display
- Remove anyOf and oneOf handling since JSON spec should only constrain
  possibilities, not expand (allowance made for enum and const to expand
  types but they show warnings when this happens).
- Avoid pushing 'null' onto available types if it already exists.
